### PR TITLE
Mannitol works on splashing and on non-failing brains, also purity matters

### PR DESCRIPTION
**Pull request by norill**
## About The Pull Request
- mannitol heals brain on splashing (fast but ineffective)
- using mannitol with syringe or dropper is twice more effective than beaker
- can now use mannitol on non-failing brains and with any amount (even <10u)
- mannitol purity matters now

## Why It's Good For The Game
Examining a moderately damaged brain tells you use mannitol, but you can only use mannitol on fully decayed brains. Trying to figure out why this doesn't work while the corpses are piling up in the medbay is kinda frustrating. This PR removes the annoying restrictions.
Also makes splashing brains with mannitol work, albeit with half the effect. On the other hand using a syringe doubles the effect. And the amount healed now takes purity into consideration.

## Changelog
:cl:
qol: Using mannitol on brains is easier now. You can use it on non-failing brains and with amounts less than 10u. Splashing now works too, but is half as effective. Use a syringe or a dropper for double effect! Also the purity of mannitol now affects the amount of damage healed.
fix: Fixed no animation and duplicated message when attacking brains
/:cl:

### CODE REVIEW

**Comment by SuperNovaa41** [comment]
This is a balance change

**Comment by norill** [comment]
which part? syringe buff? i can remove that if it's controversial

**Comment by SuperNovaa41** [comment]
> which part? syringe buff? i can remove that if it's controversial

The fact that it requires purity and the effectiveness of different applications is all balance. Its not an issue, it just needs to be labelled properly

**Comment by norill** [comment]
> 
> 
> > which part? syringe buff? i can remove that if it's controversial
> 
> The fact that it requires purity and the effectiveness of different applications is all balance. Its not an issue, it just needs to be labelled properly

it is an issue for me, i don't wanna waste gbp on balance changes for a PR that was meant to be a QoL change

**Comment by SuperNovaa41** [comment]
> > > which part? syringe buff? i can remove that if it's controversial
> > 
> > 
> > The fact that it requires purity and the effectiveness of different applications is all balance. Its not an issue, it just needs to be labelled properly
> 
> it is an issue for me, i don't wanna waste gbp on balance changes for a PR that was meant to be a QoL change

Then dont add balance changes to your pr???

File: code/modules/mob/living/brain/brain_item.dm [hunk 1]
```diff
@@ -113,24 +113,31 @@
 	if(istype(O, /obj/item/borg/apparatus/organ_storage))
 		return //Borg organ bags shouldn't be killing brains
 
-	if((organ_flags & ORGAN_FAILING) && O.is_drainable() && O.reagents.has_reagent(/datum/reagent/medicine/mannitol)) //attempt to heal the brain
+	var/healby = 2 //heals 2 damage per unit of mannitol
+	var/pour = "carefully pour"
+	var/onto = "onto"
+	if(istype(O, /obj/item/reagent_containers/syringe))
+		healby = 4 //heal 4 damage because PRECISCION
```
Comment (ATH1909) [inline code/modules/mob/living/brain/brain_item.dm:120]: true, but you're also jabbing their brain with a needle, no?

**Comment by ATH1909** [comment]
> > > > which part? syringe buff? i can remove that if it's controversial
> > > 
> > > 
> > > The fact that it requires purity and the effectiveness of different applications is all balance. Its not an issue, it just needs to be labelled properly
> > 
> > 
> > it is an issue for me, i don't wanna waste gbp on balance changes for a PR that was meant to be a QoL change
> 
> Then dont add balance changes to your pr???

But removing anything that could technically affect balance would strictly decrease the quality of this PR.

**Comment by SuperNovaa41** [comment]
> > > > > which part? syringe buff? i can remove that if it's controversial
> > > > 
> > > > 
> > > > The fact that it requires purity and the effectiveness of different applications is all balance. Its not an issue, it just needs to be labelled properly
> > > 
> > > 
> > > it is an issue for me, i don't wanna waste gbp on balance changes for a PR that was meant to be a QoL change
> > 
> > 
> > Then dont add balance changes to your pr???
> 
> But removing anything that could technically affect balance would strictly decrease the quality of this PR.

Oh yeah i like all the changes, it still needs to be labelled properly

File: code/modules/mob/living/brain/brain_item.dm (hunk 1 above)
Comment (SgtHunk) [reply code/modules/mob/living/brain/brain_item.dm:120]: It's a PRECISE needle, ATH

**Comment by zxaber** [comment]
Making mannitol work at any damage level is probably a fix to an oversight, but the effectiveness changes (and the splash effects) are balance for sure.

**Comment by norill** [comment]
fine, i'll remove the balance changes when i have time

**Comment by SuperNovaa41** [comment]
They're still good changes man...

File: code/modules/mob/living/brain/brain_item.dm [hunk 2]
```diff
@@ -113,24 +113,31 @@
...
+		pour = "precisely inject"
+		onto = "into"
+	else if(istype(O, /obj/item/reagent_containers/dropper))
+		healby = 4
+
+	if(O.reagents?.has_reagent(/datum/reagent/medicine/mannitol)) //attempt to heal the brain
```
Comment (coiax) [inline code/modules/mob/living/brain/brain_item.dm:126]: I think, with this code as written, you'd be able to "pour" any mannitol containing item onto the brain, including slices of delicious brain cake, or a cigarette injected with mannitol.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 3]
```diff
@@ -875,6 +875,22 @@
...
 	impure_chem = /datum/reagent/impurity/mannitol
 
+/datum/reagent/medicine/mannitol/expose_obj(obj/exposed_obj, reac_volume)
+	. = ..()
+	var/obj/item/organ/brain/exposed_brain = exposed_obj
+	if(!exposed_brain)
+		return
+
+	if(exposed_brain.brainmob?.health <= HEALTH_THRESHOLD_DEAD) //if the brain is fucked anyway, do nothing
+		to_chat(usr, "<span class='warning'>[exposed_brain] is far too damaged, there's nothing else we can do for it!</span>")
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm:885]: There might not be a `usr` to message, I wouldn't produce any message in this case.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 4]
```diff
@@ -875,6 +875,22 @@
...
+		return
+
+	to_chat(usr, "<span class='notice'>[exposed_brain] looks slightly better, but not much. Perhaps splashing is not a very precise method.</span>")
```
Comment (coiax) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm:888]: Make this a visible message.

File: code/modules/mob/living/brain/brain_item.dm [hunk 5]
```diff
@@ -113,24 +113,31 @@
...
 
-	if((organ_flags & ORGAN_FAILING) && O.is_drainable() && O.reagents.has_reagent(/datum/reagent/medicine/mannitol)) //attempt to heal the brain
+	var/healby = 2 //heals 2 damage per unit of mannitol
+	var/pour = "carefully pour"
+	var/onto = "onto"
+	if(istype(O, /obj/item/reagent_containers/syringe))
+		healby = 4 //heal 4 damage because PRECISCION
+		pour = "precisely inject"
+		onto = "into"
+	else if(istype(O, /obj/item/reagent_containers/dropper))
```
Comment (coiax) [inline code/modules/mob/living/brain/brain_item.dm:123]: Not sure if I really like this sort of type checking, but unsure what would be a suitable replacement.

**Review by coiax — CHANGES_REQUESTED** [review]

**Comment by coiax** [comment]
Also, the difference between Quality of Life and Balance is often just a question of perception.

I would not modify your PR in the hope of getting one specific tag or another, the accounting can be done at the end.

File: code/modules/mob/living/brain/brain_item.dm (hunk 2 above)
Comment (norill) [reply code/modules/mob/living/brain/brain_item.dm:126]: im having trouble coming up with a simple, elegant check that would differentiate between valid and invalid containers for this.
valid examples:
- beaker
- bottle
- syringe
- dropper 
- ice cup/flask/drink with mannitol pill in it
invalid examples:
- food items
- cigarettes
- pill

im thinking about going back to is_drainable, but with special case to allow droppers and syringes

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 6]
```diff
@@ -875,6 +875,19 @@
...
+	. = ..()
+	var/obj/item/organ/brain/exposed_brain = exposed_obj
+	if(!istype(exposed_brain))
+		return
+
+	exposed_brain.visible_message("<span class='notice'>[exposed_brain] looks slightly better, but not much. Perhaps splashing is not a very precise method.</span>")
+
+	var/healby = 1 //heals 1 damage per unit of mannitol
+	healby *= reac_volume
+	healby *= src.purity / REAGENT_STANDARD_PURITY
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm:888]: why src?

File: code/modules/mob/living/brain/brain_item.dm (hunk 5 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/brain/brain_item.dm:124]: there is none (currently)

File: code/modules/mob/living/brain/brain_item.dm [hunk 7]
```diff
@@ -113,24 +113,34 @@
...
+	var/healby = 2 //heals 2 damage per unit of mannitol
+	var/pour = "carefully pour"
+	var/onto = "onto"
+	var/valid_container = TRUE
+	if(istype(O, /obj/item/reagent_containers/syringe))
+		healby = 4 //heal 4 damage because PRECISCION
+		pour = "precisely inject"
+		onto = "into"
+	else if(istype(O, /obj/item/reagent_containers/dropper))
+		healby = 4
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/brain/brain_item.dm:125]: is getting 100% purity really worth 4x healing im really afraid fermichem is going to end up buffing all these items and then we find out someone already has a baby dumb way of purifying these chems.

Comment (norill) [reply code/modules/mob/living/brain/brain_item.dm:125]: why 4x? default 0.75 purity mannitol heals 2 damage per unit. with syringe/dropper it's 4 per unit, so 2x buff. and 100% purity heals 2.6 damage per unit.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/brain/brain_item.dm:125]: mental

**Pull request closed without merging**